src/lib.rs

```diff
@@ -3,30 +3,45 @@ use kvmi_sys::{kvmi_qemu2introspector, kvmi_introspector2qemu};
 use std::ffi::{CString};
 use std::ptr::{null_mut};
 use std::os::raw::{c_void, c_uchar, c_int};
+use std::sync::{Mutex, Condvar};
 
 
 #[derive(Debug)]
 pub struct KVMi {
     ctx: *mut c_void,
+    dom: *mut c_void,
+}
+
+#[derive(Debug)]
+struct KVMiCon {
+    dom: *mut c_void,
+    guard: Mutex<bool>,
+    condvar: Condvar,
 }
 
 extern "C" fn new_guest_cb(dom: *mut c_void,
                            uuid: *mut [c_uchar; 16usize],
-                           ctx: *mut c_void) -> c_int {
+                           cb_ctx: *mut c_void) -> c_int {
     println!("new guest cb !");
+    let kvmi_con = unsafe { &mut *(cb_ctx as *mut KVMiCon) };
+    let mut started = kvmi_con.guard.lock().unwrap();
+    kvmi_con.dom = dom;
+    *started = true;
+    // wake up waiters
+    kvmi_con.condvar.notify_one();
     0
 }
 
 extern "C" fn handshake_cb(arg1: *const kvmi_qemu2introspector,
                            arg2: *mut kvmi_introspector2qemu,
-                           ctx: *mut c_void) -> c_int {
+                           cb_ctx: *mut c_void) -> c_int {
     println!("handshake cb !");
     0
 }
 
 impl KVMi {
-    pub fn new() {
-        let socket_path = CString::new("/tmp/introspector").unwrap();
+    pub fn new(socket_path: String) -> KVMi {
```

Comment:
I'd suggest `AsRef<str>`
Using String you force the caller to dynamically allocate the path. `AsRef<str>` allows you to use either a `str` or a `String`

Reply:
IMHO, it would be not easy to borrow in this case. Because `socket_path` is a null-terminated `C` string while the parameter is a `Rust` string, the inverse direction (slicing `C` string) is easy.